pycid/core/macid.py

```diff
@@ -44,34 +64,243 @@ def get_all_pure_ne_in_sg(
+    ) -> List[pygambit.lib.libgambit.MixedStrategyProfile]:
+        """Uses pygambit to find the Nash equilibria of the EFG.
+        Pygambit will raise errors if solver not allowed for the game (e.g. not constant sum for lp)
+        """
+        # check if not 2 player game
+        two_player = True if len(game.players) == 2 else False
+        if solver in ["enummixed", "lcp", "lp"] and not two_player:
+            warn(
+                f"Solver {solver} not allowed for non-2 player games. Using 'simpdiv' instead."
+                "Note this will find only one NE."
+            )
+            solver = "simpdiv"
+
+        if solver == "enummixed":
+            mixed_strategies = pygambit.nash.enummixed_solve(game, rational=False)
+        elif solver == "enumpure":
+            mixed_strategies = pygambit.nash.enumpure_solve(game)
+        elif solver == "lcp":
+            mixed_strategies = pygambit.nash.lcp_solve(game, rational=False)
+        elif solver == "lp":
+            mixed_strategies = pygambit.nash.lp_solve(game, rational=False)
+        elif solver == "simpdiv":
+            mixed_strategies = pygambit.nash.simpdiv_solve(game)
+        elif solver == "ipa":
+            mixed_strategies = pygambit.nash.ipa_solve(game)
+        elif solver == "gnm":
+            mixed_strategies = pygambit.nash.gnm_solve(game)
+        else:
+            raise ValueError(f"Solver {solver} not recognised")
+        # convert to behavior strategies
+        behavior_strategies = [x.as_behavior() if solver not in ["lp", "lcp"] else x for x in mixed_strategies]
+
+        return behavior_strategies
+
+    def _behavior_to_cpd(
+        self,
+        macid: MACID,
+        state_to_infoset: Mapping[Tuple[Hashable, Tuple[Tuple[str, Any], ...]], pygambit.Infoset],
+        behavior: pygambit.lib.libgambit.MixedStrategyProfile,
+        decisions_in_sg: Union[KeysView[str], Set[str]] = None,
+    ) -> List[StochasticFunctionCPD]:
+        """Convert a behavior strategy to list of CPDs for each decision node"""
+
+        if decisions_in_sg is None:
+            decisions_in_sg = self.decisions
+
+        def _decimal_from_fraction(item: Any) -> float:
+            if isinstance(item, pygambit.Rational):
+                return float(item.numerator / item.denominator)
             else:
-                # it's an NE
-                all_pure_ne_in_sg.append(list(pp))
+                return float(item)
 
-        return all_pure_ne_in_sg
+        def _action_prob_given_parents(node: Any, **pv: Outcome) -> Mapping[str, float]:
+            """Takes the parent instantiation and outputs the prob from the infoset"""
+            pv_tuple = (macid.decision_agent[node], tuple(pv.items()))
+            # get the infoset for the node
+            infoset = state_to_infoset[pv_tuple]
+            # if the infoset does not exist, this is not a valid parent instantiation
+            # TODO check what to do here
```

Comment:
I can't actually find the reason why this way playing up before and it passes all the unittests without the if statement now so might have resolved the issue. I will keep it as it is and leave the comment just in case though because it doesn't add much overhead